nav2_behavior_tree/plugins/decorator/path_longer_on_approach.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2022 Neobotix GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <memory>
+#include <vector>
+#include "nav2_util/geometry_utils.hpp"
+
+#include "nav2_behavior_tree/plugins/decorator/path_longer_on_approach.hpp"
+
+namespace nav2_behavior_tree
+{
+
+PathLongerOnApproach::PathLongerOnApproach(
+  const std::string & name,
+  const BT::NodeConfiguration & conf)
+: BT::DecoratorNode(name, conf)
+{
+  node_ = config().blackboard->get<rclcpp::Node::SharedPtr>("node");
+}
+
+bool PathLongerOnApproach::isPathUpdated(
+  nav_msgs::msg::Path & new_path,
+  nav_msgs::msg::Path & old_path)
+{
+  return new_path != old_path && old_path.poses.size() != 0 &&
+         new_path.poses.size() != 0 &&
+         old_path.poses.back() == new_path.poses.back();
+}
+
+bool PathLongerOnApproach::isRobotInGoalProximity(
+  nav_msgs::msg::Path & old_path,
+  double & prox_leng)
+{
+  return nav2_util::geometry_utils::calculate_path_length(old_path, 0) < prox_leng;
+}
+
+bool PathLongerOnApproach::isNewPathLonger(
+  nav_msgs::msg::Path & new_path,
+  nav_msgs::msg::Path & old_path,
+  double & length_factor)
+{
+  return nav2_util::geometry_utils::calculate_path_length(new_path, 0) >
+         length_factor * nav2_util::geometry_utils::calculate_path_length(
+    old_path, 0);
+}
+
+inline BT::NodeStatus PathLongerOnApproach::tick()
+{
+  getInput("path", new_path_);
+  getInput("prox_len", prox_len_);
+  getInput("length_factor", length_factor_);
+
+  // Check if the path is updated and valid, compare the old and the new path length,
+  // given the goal proximity and check if the new path is longer
+  if (isPathUpdated(new_path_, old_path_) && isRobotInGoalProximity(old_path_, prox_len_) &&
+    isNewPathLonger(new_path_, old_path_, length_factor_) && !first_time_)
+  {
+    const BT::NodeStatus child_state = child_node_->executeTick();
+    switch (child_state) {
+      case BT::NodeStatus::RUNNING:
+        return BT::NodeStatus::RUNNING;
+      case BT::NodeStatus::SUCCESS:
+        first_time_ = true;
+        return BT::NodeStatus::SUCCESS;
+      case BT::NodeStatus::FAILURE:
+        return BT::NodeStatus::FAILURE;
+      default:
+        return BT::NodeStatus::FAILURE;
+    }
+  }
+  first_time_ = false;
```

Comment:
The last couple of things I'm wondering about are these global setters in L83/84.

We wouldn't change the `old_path_` if the `if()` condition was met to trigger the children. For the case of `RUNNING`, that has the nice result of retriggering to keep ticking without specifically storing the state of the child between runs. However, for the `SUCCESS` or `FAILURE` terminal condition, wouldn't it be necessary that the new path is ingested? 

Else, wouldn't this behavior just go in a cyclical loop that would never terminate if it actually did have to go all the way around because the obstacle in the way didn't move? Did you see that in your testing?

 `first_time_` I'm still not fully understanding what it is you're trying to accomplish with its use. Please let me know why this is around and what this is trying to solve. The check in the `if()` + change to `true` if success, and `false` at this line  created a behavior I'm not sure is sensible but I want to know what you are thinking about it. It feels to me like this should be default to `false` on initialization, then set to `true` once its been called once, and then only set back to false when the state returns to `IDLE`. Moreover, it seems like a proxy for `old_path_` essentially, where if it returns to IDLE we just clear those poses so that its empty again and then the `isPathUpdated()` check would handle it

Reply:
> We wouldn't change the old_path_ if the if() condition was met to trigger the children. For the case of RUNNING, that has the nice result of retriggering to keep ticking without specifically storing the state of the child between runs. However, for the SUCCESS or FAILURE terminal condition, wouldn't it be necessary that the new path is ingested?

We are using `ReactiveSequence` for this branch which would technically be reticking the `PathLongerOnApproach` nodes all the times. To quickly give you a short background on `ReactiveSequence`, it will tick the second child, if and only if the first child is set to SUCCESS. Also note that the first child will be reticked all the time. A nice example is given `[here](https://www.behaviortree.dev/sequencenode/)`. 

So during the state while robot is waiting, which technically here makes the BT node return a running state, remember the controller_server has been already cancelled, meaning the FollowPath is not in the RUNNING state any more. This BTNode gets reticked all the time, at this point obviously the new path gets ingested, which would eventually go into the if condition that you specify. The `FollowPath` node will be ticked if and only if the `PathLongerOnApproach` returns SUCCESS. So we have two scenarios for the SUCCESS to occur, first one is the obvious one without the obstacle. The second one is after you see the obstacle, wait for it either to see if it gets cleared, in this case as already said above, the `ReactiveSequence` makes sure that our decorator node gets reticked and compares new path (L61) with the lastly stored old path before the obstacle (L84) to see if the obstacle has cleared or not. If cleared, well and good the if condition fails and the SUCCESS is returned as seen in the L85. If not, it waits and returns the SUCCESS in L74. 

The only scenario where the FAILURE could happen is only when any of the decorator's children or grandchildren fail ( :-D )

Reply:
I do not follow this explanation. If the node ticks its children due to passing the `if()`  condition, then the `old_path_` is never set with the new value for the duration of the children ticking process. At the end when it eventually does return a SUCCESS or FAILURE, I think the `old_path_` needs to be set to the current path so that there is now a reset in behavior that we process now based on the new path information. 

Else, then if the next path comes in after this has all gone down, it will forever be stuck in that loop because `old_path_` is never updated with the new going-around-the-obstacle longer path, if eventually the result is that we **have** to take the long path around, since you check proximity with the `old_path_` that ever was updated. You would have to have an intermediary `tick()` that did not pass the `if()` statement in order for the `old_path_` to be reset, which wouldn't happen if we were still within N meters of the goal and referencing the original `old_path_` that got us into that situation in the first place.  

To me at least, it seems clear that you need to set `old_path_ = new_path_;` before all FAILURE or SUCCESS conditions in the `if()` so that it resets the execution of this behavior to be rebased on the new information that we have to go around. 

Is there something I'm misunderstanding?

Or is this where that `first_time_` thing comes in and starts to muddle things up. Why is that even there? If we simply delete the reset `first_time_ = false;` at L83 and move `old_path_ = new_path_;`  before L76, L78, and L80 (terminal conditions) doesn't that make `first_time_` actually be about only the first time and clear up my concern as well?

Reply:
I think we are getting on to the same page now. 

> If the node ticks its children due to passing the if() condition, then the old_path_ is never set with the new value for the duration of the children ticking process. 

Exactly, that is what we want when the robot cancels and waits to see if there is a new shorter path. If we update the old_path_ at the point where we wait to see if the new path is shorter, then the old path (which is longer) get's updated by the global planner (remember we are not going to cancel the planner server). This would indeed have a chain of consequences leading to the failures for the case where we check for significantly longer path lengths. 

> To me at least, it seems clear that you need to set old_path_ = new_path_; before all FAILURE or SUCCESS conditions in the if() so that it resets the execution of this behavior to be rebased on the new information that we have to go around.

This is indeed an option. I had this idea initially and decided not to do it because of the code re-use. Meanwhile I had abused the usage of `first_time_` I think. I will update the PR.